cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```
@@ -404,174 +409,231 @@ func (m *McmManager) GetMachineDeploymentSize(machinedeployment *MachineDeployme
-			if err != nil && time.Now().Before(retryDeadline) {
-				klog.Warningf("Unable to fetch Machine object %s, Error: %s , will retry in %s", machine.Name, err, conflictRetryInterval)
-				time.Sleep(conflictRetryInterval)
-				continue
-			} else if err != nil {
-				// Timeout occurred
-				klog.Errorf("Unable to fetch Machine object %s, Error: %s", machine.Name, err)
-				return err
+// resetPriorityForNotToBeDeletedMachines resets the priority of machines with nodes without ToBeDeleted taint to 3
+func (m *McmManager) resetPriorityForNotToBeDeletedMachines(mdName string) error {
+	allMachinesForMachineDeployment, err := m.getMachinesForMachineDeployment(mdName)
+	if err != nil {
+		return fmt.Errorf("unable to list all machines for node group %s, cannot proceed with scale-in of machines, Error: %+v", mdName, err)
+	}
+
+	for _, machine := range allMachinesForMachineDeployment {
+		nodeName := machine.Labels[v1alpha1.NodeLabelKey]
+		err = func() error {
+			node, err := m.nodeLister.Get(nodeName)
+			if err != nil && !kube_errors.IsNotFound(err) {
+				return fmt.Errorf("unable to get Node object %s for machine %s, cannot proceed with scale-in of machines, Error: %+v", nodeName, machine.Name, err)
+			} else if err == nil && taints.HasToBeDeletedTaint(node) {
+				// Don't update priority annotation if the taint is present on the node
+				return nil
 			}
-
-			mclone := machine.DeepCopy()
-
-			if isMachineTerminating(mclone) {
-				terminatingMachines = append(terminatingMachines, mclone)
-			} else {
-				expectedToTerminateMachineNodePairs = append(expectedToTerminateMachineNodePairs, machineNameNodeNamePair{mclone.Name, mclone.Labels["node"]})
+			val, ok := machine.Annotations[machinePriorityAnnotation]
+			if ok && val != "3" {
+				ctx, cancelFn := context.WithDeadline(context.Background(), time.Now().Add(m.maxRetryTimeout))
+				defer cancelFn()
+				for {
+					mc, err := m.machineLister.Machines(m.namespace).Get(machine.Name)
+					if err != nil {
```

Same as the another similar comment. Shouldn't you retry in case there is a failure to get the machine. The only error for which a retry should not be required is if it is a 404 and in that case one should ignore the error and continue to update the rest of the machine objects.

---

addressed in [9998790](https://github.com/gardener/autoscaler/pull/215/commits/9998790953fb68e9c9abbfb6d6140eb103a38dc5)